### DISCUSSION TITLE
How to use CodeHike with `<MDXProvider/>` (`@mdx-js/react`) in a Next.js page/layout

### DISCUSSION THREAD

**Discussion by MildTomato** [code-hike/codehike#284] [Q&A] [answered]
Hi everyone,

I'm using `<MDXProvider/>` from `@mdx-js/react` to consume the mdx source.
It seems like this is maybe a strange setup, but it means we get to have a bit more control over the layout of the page.

It seems like codehike doesn't like it though, so I was just wondering if CodeHike can be used with a set up like this?

The error looks like this - which is kind of to be expected as I guess it doesn't know what to do with the `<CH.Code>` and `<CH.Section>` components. Even if I fixed that, would it even work ?
<img width="1258" alt="Screenshot 2022-10-01 at 9 14 55 PM" src="https://user-images.githubusercontent.com/8291514/193411335-7bf560d5-1469-48b2-86fb-f4fc8cc631ee.png">

Update.. I've now added the components in via my getComponents function.
But now i'm getting the following with the <CH.Code/>

<img width="1770" alt="Screenshot 2022-10-01 at 9 46 52 PM" src="https://user-images.githubusercontent.com/8291514/193412619-d0b009da-0c96-4ceb-87ae-262e92ca016b.png">

// next.config.mjs
```js 
import nextMdx from '@next/mdx'
import remarkGfm from 'remark-gfm'
import rehypeSlug from 'rehype-slug'
import { remarkCodeHike } from '@code-hike/mdx'
import theme from 'shiki/themes/github-dark.json' assert { type: 'json' }

const withMDX = nextMdx({
  extension: /\.mdx?$/,
  options: {
    remarkPlugins: [remarkGfm, [remarkCodeHike, { theme, lineNumbers: false }]],
    rehypePlugins: [rehypeSlug],
  },

  options: {
    // This is required for `MDXProvider` component
    providerImportSource: '@mdx-js/react',
  },
})

// /** @type {NextConfig} */
const nextConfig = {
  pageExtensions: ['ts', 'tsx', 'js', 'jsx', 'md', 'mdx'],
  reactStrictMode: true,
  swcMinify: true,
}

// next.config.js
export default () => {
  const plugins = [withMDX]
  return plugins.reduce((acc, next) => next(acc), nextConfig)
}
```

// mdx file example
```

import GuideLayout from '~/layouts/guide'

export const meta = {
  author: 'Rich Haines',
  type: 'guide',
}

# Just testing out [Code Hike](https://codehike.org/)

Some normal `markdown`

## Annotation Example

\```js index.js box=1[25:39]
console.log('Some code. I am annotated!')
\```

## Focus Example

\```js next.config.js focus=1:2,7
const { remarkCodeHike } = require('@code-hike/mdx')
const theme = require('shiki/themes/monokai.json')

const withMDX = require('@next/mdx')({
  extension: /\.mdx?$/,
  options: {
    remarkPlugins: [[remarkCodeHike, { theme }]],
    rehypePlugins: [],
  },
})

module.exports = withMDX({
  pageExtensions: ['ts', 'tsx', 'js', 'jsx', 'md', 'mdx'],
})
\```

<CHSection>

## Code Links Example

We are looking at the [console.log](focus://console.js#2) function today

<CHCode>\```js console.js console.table(["Hello", "World"]) console.log("Hello World") \```</CH.Code>

</CHSection>

export default (context) => <GuideLayout meta={meta} {...context} />
```

// layout
```js
import { MDXProvider } from '@mdx-js/react'
import getComponents, { MDXComponentsType } from '~/lib/getComponents'

const ContentsLayout = (props) => {
  return (
    <>
      <MDXProvider components={getComponents()}>
        <div className="prose mx-auto my-32">{props.children}</div>
      </MDXProvider>
    </>
  )
}

export default ContentsLayout
```

// components for MDXprovider
```js
import NumberSection from '~/components/NumberSection'
import TestDiagram from '~/diagrams/testdiagram'
import { Section as CHSection, Code as CHCode } from '@code-hike/mdx/components'

export default function mdxComponents() {
  const components = {
    NumberSection: (props: any) => {
      return <NumberSection {...props} />
    },
    TestDiagram: (props: any) => {
      return <TestDiagram />
    },
   CHSection: (props: any) => {
      return <CHSection {...props} />
    },
    CHCode: (props: any) => {
      return <CHCode {...props} />
    },
  }

  return components
}
```

**Comment by pomber** [comment, marked as answer]
Try this https://codehike.org/docs/configuration#auto-import

(From my phone, didn't read the entire issue 😁)

**Reply by MildTomato** [reply]
Awesome, trying it out now

**Reply by MildTomato** [reply]
Weird, I'm still getting the `TypeError: Cannot read properties of undefined (reading 'length')`. I'll keep trying though

**Reply by pomber** [reply]
Make sure to pass `CH` in the components objects. Something like:

```
import { CH } from "@code-hike/mdx/components"

export default function mdxComponents() {
  const components = {
    NumberSection: (props: any) => {
      return <NumberSection {...props} />
    },
    TestDiagram: (props: any) => {
      return <TestDiagram />
    },
   CH
  }

  return components
}
```

If it still doesn't work, you can create a small reproduction repo and I'll take a look later.

**Reply by MildTomato** [reply]
hmm i still get the same error.
I'll publish a reproduction repo as you suggest 👍 - i'll paste a link in here

**Reply by MildTomato** [reply]
Here we go. [Reproduction repo](https://github.com/MildTomato/code-hike-nextjs-mdx).

```
npm i
npm run dev
```

go to localhost:3000
then the mdx file that will break is `localhost:3000/code-hike`
page for that can be found at `./pages/code-hike.mdx`

**Reply by pomber** [reply]
can you make it smaller? start removing things until it starts working or you can't remove anything else, so it's easier to debug

**Reply by MildTomato** [reply]
sure, cleaned it up 😄

**Reply by MildTomato** [reply]
If I remove the `CH` component from the components injected into MDXProvider it works 😂 
CH.Section seems to render fine though, so if you remove CH.Code the markdown will then render

**Reply by pomber** [reply]
I think the problem is in your `next.config.mjs`, you are defining `options` two times. After you fix that, I think you can remove the `autoImport: false`, for your case it should work without manually importing `CH`.

**Reply by MildTomato** [reply]
🙈 I'm so sorry - it's exactly that!

how did i not see that?!
